src/target/riscv/riscv.c

```diff
@@ -1601,18 +1602,29 @@ static int riscv_trigger_detect_hit_bits(struct target *target, int64_t *unique_
 				break;
 			case CSR_TDATA1_TYPE_MCONTROL:
 				hit_mask = CSR_MCONTROL_HIT;
+				if (get_field(tdata1, CSR_MCONTROL_TIMING) == CSR_MCONTROL_TIMING_AFTER)
+					r->need_single_step = false;
 				break;
 			case CSR_TDATA1_TYPE_MCONTROL6:
 				hit_mask = CSR_MCONTROL6_HIT0 | CSR_MCONTROL6_HIT1;
+				int hit0 = get_field(tdata1, CSR_MCONTROL6_HIT0);
+				int hit1 = get_field(tdata1, CSR_MCONTROL6_HIT1);
+				int trigger_retired_info = (hit1 << 1) | hit0;
+				if (trigger_retired_info == CSR_MCONTROL6_HIT0_AFTER
+				    || trigger_retired_info == CSR_MCONTROL6_HIT0_IMMEDIATELY_AFTER)
+					r->need_single_step = false;
 				break;
 			case CSR_TDATA1_TYPE_ICOUNT:
 				hit_mask = CSR_ICOUNT_HIT;
+				r->need_single_step = false;
```

Comment:
`r->need_single_step` should be reset only if `tdata1 & hit_mask` is not zero.

With current implementation in case you have a `mcontrol` trigger at `tselect == 0` and `icount` trigger at `tselect == 1` and  `mcontrol` is hit but the timing is not `CSR_MCONTROL_TIMING_AFTER` `r->need_single_step` will be `false`.

Reply:
yeah, next patch considering this scenario, avoiding `icount/itrigger/etrigger` overwrites `mcontrol.timing/mcontrol6.hit0[1]`  is `before` situation.